### PR TITLE
Allow usage of datastore encryption with a pregenerated key

### PR DESCRIPTION
**Pull request by AngryDeveloper**
Closes https://github.com/StackStorm/stackstorm-ha/issues/9

This will allow the deployment to use datastore encryption if we provide
the required key through a K8S secret.

The secret's name and the key containing the effective json is configurable through values.yaml

By default the feature is disabled (and the chart behavior is unchanged).

The secret is mounted in the following deployments:

- st2api
- st2actionrunner
- st2sensorcontainers
- st2workflowengine
- ~st2rulesengine~

I might have missed some where the key is a dependency but I didn't run into any issue during my tests.

### CODE REVIEW

**Comment by armab** [comment]
Closes https://github.com/StackStorm/stackstorm-ha/issues/9

File: values.yaml [hunk 1]
```diff
@@ -62,6 +62,17 @@ st2:
...
+  # Optional configuration for handling a precomputed datastore encryption key
+  datastoreEncryption:
+    enabled: false
+    # Where will the key be mounted in the containers
+    encryptionKeyPath: /etc/st2/keys
+  # secret:
+  #   # The name of the secret in which the key is stored
+  #   name: st2-encryption-key
+  #   # The key containing the json formated encryption key
+  #   key: datastore_encryption_key.json
```
Comment (armab) [inline values.yaml]: For simplicity reasons, I'd prefer only the following Helm value to be exposed to the user:

```yaml
secrets:
  st2:
    # ST2 crypto key for the  K/V datastore. 
    # See https://docs.stackstorm.com/datastore.html#securing-secrets-admin-only for more info.
    # Warning! Replace with your own generated key!
    #datastore_crypto_key: {"hmacKey": {"hmacKeyString": "", "size": 256}, "size": 256, "aesKeyString": "", "mode": "CBC"}    
```

And so if `secrets.st2.datastore_crypto_key` value is present, st2 crypto key for the datastore will be enabled automatically with all the K8s secrets under the hood and defaults like https://github.com/StackStorm/st2-packages/blob/e86adb5e75b5013dac43198f9e464b29cb08ed58/scripts/st2bootstrap-deb.sh#L335-L361


See for example other secrets like https://github.com/StackStorm/stackstorm-ha/blob/master/templates/secrets_ssh.yaml for more mechanics.

With that, crypto key path would be `/etc/st2/keys/datastore_key.json`.
Services that need shared access are `st2api st2sensorcontainer st2workflowengine st2actionrunner` and `st2client` pods.

Comment (AngryDeveloper) [reply values.yaml]: I've implemented as you advised, it's way less complex for the end user indeed.

**Comment by AngryDeveloper** [comment]
> 
> 
> We'll also need checksum annotations referencing new secret.
> 
> See
> 
> https://github.com/StackStorm/stackstorm-ha/blob/d59b8c4d9ea133fa0283c975b0053360381c153e/templates/deployments.yaml#L998-L1000
> 
> for example

Will implement as soon as possible.

File: templates/deployments.yaml [hunk 2]
```diff
@@ -598,12 +633,25 @@ spec:
...
         - name: st2-config-vol
           configMap:
             name: {{ .Release.Name }}-st2-config
+        {{- if .Values.secrets.st2.datastore_crypto_key }}
+        - name: st2-encryption-key-vol
+          secret:
+            secretName: {{ .Release.Name }}-st2-datastore-crypto-key
+            items:
+            - key: datastore_crypto_key
+              path: encryption_key.json
```
Comment (armab) [inline templates/deployments.yaml]:
```suggestion
              path: datastore_key.json
```

File: templates/deployments.yaml [hunk 3]
```diff
@@ -447,12 +466,25 @@ spec:
         - name: st2-config-vol
           mountPath: /etc/st2/st2.user.conf
           subPath: st2.user.conf
+        {{- if .Values.secrets.st2.datastore_crypto_key }}
```
Comment (armab) [inline templates/deployments.yaml]: Following the https://github.com/StackStorm/st2-packages/blob/e86adb5e75b5013dac43198f9e464b29cb08ed58/scripts/st2bootstrap-deb.sh#L356-L360 I don't think `st2rulesengine` needs access to st2 datastore crypto key.

@m4dcoder can you confirm?

File: templates/deployments.yaml [hunk 4]
```diff
@@ -831,6 +879,8 @@ spec:
       annotations:
         checksum/config: {{ include (print $.Template.BasePath "/configmaps_st2-conf.yaml") $ | sha256sum }}
         checksum/packs: {{ include (print $.Template.BasePath "/configmaps_packs.yaml") $ | sha256sum }}
+        {{- if $.Values.secrets.st2.datastore_crypto_key }}
+        checksum/datastore-key: {{ include (print $.Template.BasePath "/secrets_datastore_crypto_key.yaml") & | sha256sum }}
```
Comment (armab) [inline templates/deployments.yaml:883]: `Helm-lint` is unhappy about this one: https://app.circleci.com/pipelines/github/StackStorm/stackstorm-ha/570/workflows/86ddf92e-4929-4456-8cac-0c76d0085ee0/jobs/2231/steps

```suggestion
        checksum/datastore-key: {{ include (print $.Template.BasePath "/secrets_datastore_crypto_key.yaml") $ | sha256sum }}
```
Looks like it's a typo here.

File: templates/deployments.yaml (hunk 3 above)
Comment (armab) [reply templates/deployments.yaml:481]: `st2rulesengine` doesn't need access to st2 datastory crypto key. Please remove it.

File: CHANGELOG.md [hunk 5]
```diff
@@ -27,6 +28,7 @@
 * Add an option to inject hostAliases in the st2actionrunner containers (#114)
 * Add support for Service Accounts (#117) (by @Vince-Chenal)
 
+
```
Comment (armab) [inline CHANGELOG.md:31]: To minimize the diff
```suggestion
```

File: templates/deployments.yaml [hunk 6]
```diff
@@ -831,6 +871,8 @@ spec:
       annotations:
         checksum/config: {{ include (print $.Template.BasePath "/configmaps_st2-conf.yaml") $ | sha256sum }}
         checksum/packs: {{ include (print $.Template.BasePath "/configmaps_packs.yaml") $ | sha256sum }}
+        {{- if $.Values.secrets.st2.datastore_crypto_key }}
```
Comment (armab) [inline templates/deployments.yaml:874]: missing a closing `end` tag for `if`?

File: templates/deployments.yaml [hunk 7]
```diff
@@ -979,6 +1034,8 @@ spec:
       annotations:
         checksum/config: {{ include (print $.Template.BasePath "/configmaps_st2-conf.yaml") . | sha256sum }}
         checksum/ssh: {{ include (print $.Template.BasePath "/secrets_ssh.yaml") . | sha256sum }}
+        {{- if $.Values.secrets.st2.datastore_crypto_key }}
```
Comment (armab) [inline templates/deployments.yaml:1037]: same with the closing `if`

File: templates/deployments.yaml (hunk 6 above)
Comment (armab) [reply templates/deployments.yaml:874]: In fact, do we really need `if` -> `end` for the datastore crypto key checksum in annotations?
I mean even if `secrets_datastore_crypto_key.yaml` is empty if no key is provided, when it's populated the checksum is changed.

Comment (AngryDeveloper) [reply templates/deployments.yaml:874]: Yep, not quite awake yet...

Indeed, removing the conditional altogether would be cleaner, implementing.

**Pull request merged**